PSG-1727 pipeline

jurajseffer opened the pull request:
## Before merge:
- [x] Amend repo settings to not require CircleCI jobs anymore
- [ ] Put `PSG-1727` in the commit message when merging

Changes:
- builds images in parallel using matrix and always push them
- pushes YAML config files into S3 on `dev` and `main` under hash and {branch}_latest path to be then consumed by backend
- removed CircleCI config
- added YAML configs linting
- use `https` for cloning for the public repo instead of `ssh` because it doesn't require SSH key auth - can be run from anywhere

pdp10 commented on .github/workflows/build_image.yaml:
```diff
@@ -0,0 +1,72 @@
+          IMAGE_NAME="${{ inputs.name }}:${VERSION}"
+          ECR_IMAGE_NAME="${ECR_PATH}/${IMAGE_NAME}"
+          ECR_BRANCH_LATEST_NAME="${ECR_PATH}/${{ inputs.name }}:${BRANCH_NAME}_latest"
+          PATHOGEN_ARG=""
+          if [[ ! -z "${{ inputs.pathogen }}" ]]; then
+            PATHOGEN_ARG="--build-arg pathogen=${{ inputs.pathogen }}"
+          fi
```
This has changed from the circleci config. All images receive the pathogen, so that we can have pathogen specific dependencies (e.g. fastqc for sars-cov-2-pipeline could have a different version from fastqc for tb-pipeline)

jurajseffer replied:
All jobs supply `pathogen` in the matrix but since they all do, I'll remove this logic and always supply the build arg.

pdp10 reviewed (commented):
It would be good if @cramshaw  can have a look as I am not an expert in github actions.

@jurajseffer
1. Regarding dev/main as only branches to push the images, this mean that we might be in a situation in which we cannot apply a quick fix to the pipeline which we don't want yet in dev (e.g. for testing only). This has happened a few times - the last time was for the QC_STATUS field in the synthetic-pipeline, just before the release.
2. I remember in the past you suggested to run the pipeline via minikube before pushing the images. I cannot see that code in here, so I wonder when/ if you plan to implement that.
3. I recently made some minor changes to the pipeline circleci config : https://github.com/Congenica/psga-pipeline/pull/168/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47 but it should be easy to port them to github actions.

jurajseffer commented:
1. Amended to always push on a branch build
2. No testing implementation here yet, that's scheduled for later
3. Updated to run git submodules - also updated to use `https` instead of `ssh` so that we don't need private key to auth even though the repo is public
4. Will do

pdp10 commented on .gitmodules:
```diff
@@ -1,3 +1,3 @@
 [submodule "docker/sars_cov_2/ncov2019-artic-nf"]
 	path = docker/sars_cov_2/ncov2019-artic-nf
-	url = git@github.com:Congenica/ncov2019-artic-nf.git
```
Thank you

cramshaw commented on .github/workflows/publish_configs.yaml:
```diff
@@ -0,0 +1,33 @@
+        run: |
+          for version in $(echo ${{ inputs.versions }} | sed "s/,/ /g"); do
+            echo "Publishing version $version"
+            aws s3 cp psga/config/psga-schema-sars-cov-2.yaml ${{ secrets.S3_BUCKET_PATH }}/${version}/psga-schema-sars-cov-2.yaml
```
I would like this to be dynamic rather than having to identify each one manually. Hard coded like this will do for a first version though

jurajseffer replied:
It now lists over the files. I wasn't sure whether the configs folder will have all sorts of random things in it.

cramshaw commented on .github/workflows/pipeline.yaml:
```diff
@@ -0,0 +1,76 @@
+      name: ${{ matrix.image.name }}
+      dockerfile: ${{ matrix.image.dockerfile }}
+      pathogen: ${{ matrix.image.pathogen }}
+  publish_configs_dev:
```
As I believe Piero has pointed out - it would be good to push images for feature branches as well, as it allows us to test them out in dev before merging. Can we add that please?

jurajseffer replied:
It now pushes images and configs from branches too.

cramshaw reviewed (approved):
I'm happy to proceed with this, but I do think that pushing images for feature branches would be a nice addition.
